proxy/src/main/java/com/wavefront/agent/ProxyUtil.java

```diff
@@ -0,0 +1,189 @@
+    }
+  }
+
+  /**
+   * Read or create proxy id for this machine. Reads the UUID from specified file,
+   * or from ~/.dshell/id if idFileName is null.
+   *
+   * @param idFileName file name to read proxy ID from.
+   * @return proxy id
+   */
+  static UUID getOrCreateProxyIdFromFile(@Nullable String idFileName) {
+    File proxyIdFile;
+    UUID proxyId = UUID.randomUUID();
+    if (idFileName != null) {
+      proxyIdFile = new File(idFileName);
+    } else {
+      File userHome = new File(System.getProperty("user.home"));
+      if (!userHome.exists() || !userHome.isDirectory()) {
+        logger.severe("Cannot read from user.home, quitting");
+        System.exit(1);
+      }
+      File configDirectory = new File(userHome, ".dshell");
+      if (configDirectory.exists()) {
+        if (!configDirectory.isDirectory()) {
+          logger.severe(configDirectory + " must be a directory!");
+          System.exit(1);
+        }
+      } else {
+        if (!configDirectory.mkdir()) {
+          logger.severe("Cannot create .dshell directory under " + userHome);
+          System.exit(1);
+        }
+      }
+      proxyIdFile = new File(configDirectory, "id");
+    }
+    if (proxyIdFile.exists()) {
+      if (proxyIdFile.isFile()) {
+        try {
+          proxyId = UUID.fromString(Objects.requireNonNull(Files.asCharSource(proxyIdFile,
+              Charsets.UTF_8).readFirstLine()));
+          logger.info("Proxy Id read from file: " + proxyId);
+        } catch (IllegalArgumentException ex) {
+          logger.severe("Cannot read proxy id from " + proxyIdFile +
+              ", content is malformed");
+          System.exit(1);
+        } catch (IOException e) {
+          logger.log(Level.SEVERE, "Cannot read from " + proxyIdFile, e);
+          System.exit(1);
+        }
+      } else {
+        logger.severe(proxyIdFile + " is not a file!");
+        System.exit(1);
+      }
+    } else {
+      logger.info("Proxy Id created: " + proxyId);
+      try {
+        Files.asCharSink(proxyIdFile, Charsets.UTF_8).write(proxyId.toString());
+      } catch (IOException e) {
+        logger.severe("Cannot write to " + proxyIdFile);
+        System.exit(1);
+      }
+    }
+    return proxyId;
+  }
+
+  /**
+   * Return a unique process identifier used to prevent collisions in ~proxy metrics.
+   * Try to extract system PID from RuntimeMXBean name string (usually in the
+   * "11111@hostname" format). If it's not parsable or an extracted PID is too low,
+   * for example, when running in a containerized environment, chances of ID collision
+   * are much higher, so we use a random 32bit hex string instead.
+   *
+   * @return unique process identifier string
+   */
+  static String getProcessId() {
+    try {
+      final String runtime = ManagementFactory.getRuntimeMXBean().getName();
+      if (runtime.indexOf("@") >= 1) {
+        long id = Long.parseLong(runtime.substring(0, runtime.indexOf("@")));
+        if (id > 1000) {
```

Comment:
1000 is arbitrary, since pids are assigned based on request, ofc in a containerized environment the pid will almost always be 1 as the proxy will be the first and basically only thing running on the container (or should be).  Other environments it's a mixed bag based on the OS and what is running.  If collisions are a problem maybe the pid isn't best choice and best choice is just the random hex below.  1 and 0 are reserved (most of the time) and anything after that is fair game, though most of the low numbers will be consumed by low level kernel processes and other system services that start well before other services.

Reply:
useless trivia fact: in container envs with our proxy image the pid is almost always 15 (occasionally it's off by 1).
but in general i agree, let's stick to random hex.